### ISSUE TITLE
Missing indexes for some SELECT queries, which are frequently executed by CC

### ISSUE THREAD

**Issue by krumts** [cloudfoundry/cloud_controller_ng#848]
## Issue

Several of the queries which Cloud Controller executes very frequently (at least in our CF setup) run with table scans, without an index. The result is increased CPU usage on the DB and suboptimal performance. 

## Context

We are operating a CF system (currently on rel. 263) and I had there the following experience:
We recently had to scale the number of API nodes, as performance was suffering - we had only two API instances (i.e. had not followed the sizing recommendations).
We increased the number of API nodes, and are now at 10 API nodes, which brought us some performance improvement.
However, as we did the change, the CPU load on the DB (an RDS Postgres instance on AWS) went from about 40% at peaks to about 80%. 
While the bottleneck we had was definitely the too few CC nodes, it looks like we are moving to a new bottleneck on the DB, so that adding new API nodes would not really help.

Therefore, I spent some time looking at the statement statistics, to figure out what are the most expensive ones.

I reset the statement statistics in Postgres and one working day later the top statements I see are these:

```
datname     |  calls   |    total_time     |   rows    |                                                                                                       substring                                    
                                                               
----------------+----------+-------------------+-----------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
ccdb           |  3545679 |  40664123.0599971 |   2382551 | SELECT "builds".* FROM "builds" INNER JOIN "apps" ON ("apps"."guid" = "builds"."app_guid") WHERE ("apps"."guid" = 'cc633980-9632-434c-8624-fe618ea6eb21') ORDER BY "created_at" DESC, "id" DESC LIMIT 1
ccdb           |   324166 |  37308677.2619999 |  16207535 | SELECT * FROM "app_usage_events" ORDER BY "id" ASC LIMIT ? OFFSET ?
ccdb           |   324166 |  24637331.5350002 |    324166 | SELECT count(*) AS "count" FROM "app_usage_events" LIMIT ?
ccdb           |   536826 |  22978274.0120005 |  26783042 | SELECT * FROM "service_usage_events" WHERE (service_guid  IN  ('ad14ecbe-ed72-4e1a-a25e-5c49c79b9d05', 'bc3690b2-cc50-4475-b2cf-44d68c51f9d3', 'c4ec80ae-5949-4a79-99ef-e42a894d73d3', '3939eedf-1a36-49
ccdb           |   536825 |  17283755.9309999 |    536825 | SELECT count(*) AS "count" FROM "service_usage_events" WHERE (service_guid  IN  (?, ?, ?, ?, ?)) LIMIT ?
ccdb           |   992378 |  11299723.8330005 |    992378 | SELECT count(*) AS "count" FROM "service_bindings" WHERE ("service_bindings"."service_instance_guid" = ?) LIMIT ?
ccdb           |   899742 |  10004730.4900005 |   1594664 | SELECT * FROM "service_bindings" WHERE ("service_bindings"."app_guid" = ?)
ccdb           |   796375 |  8873692.55499966 |    796375 | SELECT count(*) AS "count" FROM "service_bindings" WHERE ("service_bindings"."app_guid" = ?) LIMIT ?
```

1) The first one (select from builds), seems to be very similar to the ones described in Issue #783 (on droplets and packages).
It runs without an index and scans through the 30.000+ entries (currently in our system). I've seen times ranging from 13 to 25 millis, which is much for a simple query.
I am pretty sure an index on the app_guid would help, as it did for the packages and droplets tables.
```
ccdb=> explain analyze SELECT "builds".* FROM "builds" INNER JOIN "apps" ON ("apps"."guid" = "builds"."app_guid") WHERE ("apps"."guid" = 'cc633980-9632-434c-8624-fe618ea6eb21') ORDER BY "created_at" DESC, "id" DESC LIMIT 1
;
                                                                   QUERY PLAN                                                                    
-------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=1493.78..1493.78 rows=1 width=264) (actual time=13.397..13.398 rows=1 loops=1)
   ->  Sort  (cost=1493.78..1493.79 rows=5 width=264) (actual time=13.394..13.394 rows=1 loops=1)
         Sort Key: builds.created_at, builds.id
         Sort Method: top-N heapsort  Memory: 25kB
         ->  Nested Loop  (cost=0.29..1493.75 rows=5 width=264) (actual time=0.287..13.330 rows=20 loops=1)
               ->  Index Only Scan using apps_v3_guid_index on apps  (cost=0.29..8.30 rows=1 width=37) (actual time=0.045..0.049 rows=1 loops=1)
                     Index Cond: (guid = 'cc633980-9632-434c-8624-fe618ea6eb21'::text)
                     Heap Fetches: 0
               ->  Seq Scan on builds  (cost=0.00..1485.40 rows=5 width=264) (actual time=0.236..13.209 rows=20 loops=1)
                     Filter: (app_guid = 'cc633980-9632-434c-8624-fe618ea6eb21'::text)
                     Rows Removed by Filter: 33685
 Planning time: 0.136 ms
 Execution time: 13.442 ms
 ```

 2) There are 3 queries among the top ones, which select data from service_bindings. All of them are also running a table scan (ranging in our case ~10 to ~15 millis), and could be optimized by adding indexes on the service_instance_guid and app_guid columns.

filter on app_guid:
 ```
 ccdb=> explain analyze SELECT * FROM "service_bindings" WHERE ("service_bindings"."app_guid" = '88e0d271-7733-4ad7-8753-234322e74408');
                                                    QUERY PLAN                                                    
 -----------------------------------------------------------------------------------------------------------------
  Seq Scan on service_bindings  (cost=0.00..2516.01 rows=2 width=1857) (actual time=4.003..10.518 rows=2 loops=1)
    Filter: (app_guid = '88e0d271-7733-4ad7-8753-234322e74408'::text)
    Rows Removed by Filter: 23352
  Planning time: 0.056 ms
  Execution time: 10.547 ms
 (5 rows)

 ```
filter on service_instance_guid:
 ```
 ccdb=> explain analyze SELECT count(*) AS "count" FROM "service_bindings" WHERE ("service_bindings"."service_instance_guid" = 'fff31097-c819-45e8-9daf-e87a9a876173') LIMIT 5;
                                                        QUERY PLAN                                                        
 -------------------------------------------------------------------------------------------------------------------------
  Limit  (cost=2516.02..2516.03 rows=1 width=0) (actual time=9.248..9.251 rows=1 loops=1)
    ->  Aggregate  (cost=2516.02..2516.03 rows=1 width=0) (actual time=9.245..9.246 rows=1 loops=1)
          ->  Seq Scan on service_bindings  (cost=0.00..2516.01 rows=2 width=0) (actual time=8.012..9.237 rows=1 loops=1)
                Filter: (service_instance_guid = 'fff31097-c819-45e8-9daf-e87a9a876173'::text)
                Rows Removed by Filter: 23363
  Planning time: 0.070 ms
  Execution time: 9.283 ms
 (7 rows)

 ```

 3) There are two expensive statements, doing table scans on service_usage_events filtering on the service_guid. Adding an index on service_guid should help.
 ```
 ccdb=> explain analyze SELECT count(*) AS "count" FROM "service_usage_events" WHERE (service_guid  IN  ('ad14ecbe-ed72-4e1a-a25e-5c49c79b9d05', 'bc3690b2-cc50-4475-b2cf-44d68c51f9d3', 'c4ec80ae-5949-4a79-99ef-e42a894d73d3', '91534880-854a-4340-8c1e-af4292470c0f')) LIMIT 5;
                                                                                                  QUERY PLAN                                                                                                  
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=5080.61..5080.62 rows=1 width=0) (actual time=46.511..46.514 rows=1 loops=1)
   ->  Aggregate  (cost=5080.61..5080.62 rows=1 width=0) (actual time=46.508..46.509 rows=1 loops=1)
         ->  Seq Scan on service_usage_events  (cost=0.00..5059.05 rows=8625 width=0) (actual time=0.014..36.690 rows=9108 loops=1)
               Filter: (service_guid = ANY ('{ad14ecbe-ed72-4e1a-a25e-5c49c79b9d05,bc3690b2-cc50-4475-b2cf-44d68c51f9d3,c4ec80ae-5949-4a79-99ef-e42a894d73d3,91534880-854a-4340-8c1e-af4292470c0f}'::text[]))
               Rows Removed by Filter: 67586
 Planning time: 0.079 ms
 Execution time: 46.558 ms
(7 rows)
```

## Steps to Reproduce

Can't really pin-point at the moment which API calls cause which of the queries listed above. 
If needed, I could try to do this analysis.

## Expected result

I would expect that most/all queries which are called very frequently (I know this is pretty imprecise) are using indexes and not full table scans.

## Current result

See above in "context" - there is a number of queries always doing a full table scan, resulting in suboptimal performance

## Possible Fix
Will it be possible to get the following indexes:
1) builds (app_guid)
2) service_bindings (app_guid)
   service_bindings (service_instance_guid)
3) service_usage_events (service_guid)

I could also provide (separate) pull requests with the migration scripts (looking how it was done in https://github.com/cloudfoundry/cloud_controller_ng/commit/485dcc6a4789af71952e42304eb09f2b017f09d5). 

I am pretty sure these changes will let the DB process much more queries with the same resources. Shall I submit pull requests, or is db migration something you'd like to do on your own (assuming you agree to the suggested changes)?

## P.S
For the other two statements among the top ones:
```
ccdb           |   324166 |  37308677.2619999 |  16207535 | SELECT * FROM "app_usage_events" ORDER BY "id" ASC LIMIT ? OFFSET ?
ccdb           |   324166 |  24637331.5350002 |    324166 | SELECT count(*) AS "count" FROM "app_usage_events" LIMIT ?
```
I'd like to post a separate issue and discuss if / what could be optimized there. Therefore no further comments here.

**Comment by zrob** [comment]
@krumts thanks for the detailed report.  i'll get this scheduled for our community pair to take a look at.

**Comment by tcdowney** [comment]
Thanks for the report @krumpts.  Just for reference, I believe these are the endpoints that are creating queries that are missing the indices:

* `builds` -- there are many places where call `app.builds` to get the builds for a given app
* [service_bindings_controller](https://github.com/cloudfoundry/cloud_controller_ng/blob/6a2b976565d8dca14e111831ab12c4d98f05ae32/app/controllers/services/service_bindings_controller.rb#L155) -- users can query on `app_guid` and `service_instance_guid`
* [service_usage_events_controller](https://github.com/cloudfoundry/cloud_controller_ng/blob/6a2b976565d8dca14e111831ab12c4d98f05ae32/app/controllers/services/service_usage_events_controller.rb#L5) -- users can query on `service_guid`

**Comment by tcdowney** [comment]
Hey @krumts, we added the missing indexes in b44ae3cb312841b7dd8f8b3cf39eab6fe95ede55 and also added an index to  `service_instance_type` for `service_usage_events` since users can query on that column via the api.  Once these go through the pipes and make it through to a CF release you should see some performance improvements. 👍

**Comment by krumts** [comment]
@tcdowney thanks a lot!
I'll track when the change enters a CF release and can give you feedback as soon as we have it in production.
Thanks!

**Comment by zrob** [comment]
@krumts closing the issue for now, would love to hear if you do or do not see improvements when you get this deployed

**Comment by krumts** [comment]
@tcdowney @zrob sorry for the late reply, I was on vacation for a while.
Meanwhile we updated our system to cf-release 270, and I can now confirm that the changes had a good visible positive effect.
We had once more scaled the DB instance (before updating to cf 270) because the situation got worse. After this scale up, the CPU usage peaks got at ~ 50% (from over 90%).
The update to cf-release 270 reduced the CPU usage significantly, and now the peaks are at about 25%.
I also verified the in the DB, that the statements are now executed with index searches and not with full table scans.
Thanks a lot for the good improvement!
